presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushAggregationIntoTableScan.java

```diff
@@ -0,0 +1,225 @@
+
+        List<ColumnHandle> groupByColumns = groupingSets.getGroupingKeys().stream()
+                .map(groupByColumn -> assignments.get(groupByColumn.getName()))
+                .collect(toImmutableList());
+
+        Optional<AggregationApplicationResult<TableHandle>> aggregationPushdownResult = metadata.applyAggregation(
+                context.getSession(),
+                tableScan.getTable(),
+                aggregateFunctions,
+                assignments,
+                ImmutableList.of(groupByColumns));
+
+        if (aggregationPushdownResult.isEmpty()) {
+            return Result.empty();
+        }
+
+        AggregationApplicationResult<TableHandle> result = aggregationPushdownResult.get();
+
+        // The new scan outputs should be the symbols associated with grouping columns plus the symbols associated with aggregations.
+        ImmutableList.Builder<Symbol> newScanOutputs = new ImmutableList.Builder<>();
+        newScanOutputs.addAll(tableScan.getOutputSymbols());
+
+        ImmutableBiMap.Builder<Symbol, ColumnHandle> newScanAssignments = new ImmutableBiMap.Builder<>();
+        newScanAssignments.putAll(tableScan.getAssignments());
+
+        Map<String, Symbol> variableMappings = new HashMap<>();
+
+        for (Assignment assignment : result.getAssignments()) {
+            Symbol symbol = context.getSymbolAllocator().newSymbol(assignment.getVariable(), assignment.getType());
+
+            newScanOutputs.add(symbol);
+            newScanAssignments.put(symbol, assignment.getColumn());
+            variableMappings.put(assignment.getVariable(), symbol);
+        }
+
+        List<Expression> newProjections = result.getProjections().stream()
+                .map(expression -> ConnectorExpressionTranslator.translate(expression, variableMappings, new LiteralEncoder(metadata)))
+                .collect(toImmutableList());
+
+        verify(aggregationOutputSymbols.size() == newProjections.size());
+
+        Assignments.Builder assignmentBuilder = Assignments.builder();
+        IntStream.range(0, aggregationOutputSymbols.size())
+                .forEach(index -> assignmentBuilder.put(aggregationOutputSymbols.get(index), newProjections.get(index)));
+
+        ImmutableBiMap<Symbol, ColumnHandle> scanAssignments = newScanAssignments.build();
+        ImmutableBiMap<ColumnHandle, Symbol> columnHandleToSymbol = scanAssignments.inverse();
+        // projections assignmentBuilder should have both agg and group by so we add all the group bys as symbol references
+        groupingSets.getGroupingKeys()
+                .forEach(groupBySymbol -> {
+                    // if the connector returned a new mapping from oldColumnHandle to newColumnHandle, groupBy needs to point to
+                    // new columnHandle's symbol reference, otherwise it will continue pointing at oldColumnHandle.
+                    ColumnHandle originalColumnHandle = assignments.get(groupBySymbol);
```

Comment:
Intellij warns me

> 'Map<String, ColumnHandle>' may not contain keys of type 'Symbol' 

and indeed i get NPE in the next line, as `originalColumnHandle` comes out as null

Reply:
updated the code so it now does `assignments.get(groupBySymbol.getName());` so there is no warning anymore. I think it is a valid assumption though that the groupBy Symbol name must be present in original scan assignment. If you think that is a wrong assumption I can add a defensive check and return an empty plan if the defensive check fails.